Panache Support multiple persistence unit in Hibernate Reactive

lucamolteni opened the pull request:
* Integration test for multiple reactive persistence units and Panache
* Backported from ORM the handling of different persistence units in entity in Panache
* withSession overload to take the PU name
* Added Panache.withTransaction overload to select persistence unit
* Execute update without entity runs on default session
* Use ComputingCache instead of a Map<String, LazyValue>
* Keep track of onDemand created sessions to close them accordingly
* create map of entity => PU in kotlin as well (identical to what happens in Hibernate ORM)
* Aligned reactive blocking to orm repositories by using @GenerateBridge to call getSession
* getSession shouldn't be static to be substituted by @GenerateBridge
* Updated Panache documentation

# Release Notes: Hibernate Reactive Panache Enhancements

Complete support for multiple persistence units in Hibernate Reactive Panache, bringing feature parity with Hibernate ORM
Panache.

## New API Methods:

```java

// Specify persistence unit for sessions
@WithSession("secondary")

or

Panache.withSession("secondary", () -> { ... })

// Specify persistence unit for transactions  
@WithTransaction("secondary")

or

Panache.withTransaction("secondary", () -> { ... })

// Repository operations work with entity's configured PU
secondaryRepository.findAll() // Uses "secondary" PU automatically
```

## Kotlin Module Independence

The Hibernate Reactive Panache Kotlin module has been changed to depend only on the `hibernate-reactive-panache-common` module, aligning it to the ORM version.
Backwards compatibility is maintained through a copy of `Panache` class inside the Kotlin module


```java
// New primary class for Kotlin users
io.quarkus.hibernate.reactive.panache.kotlin.Panache

// Deprecated compatibility class  
io.quarkus.hibernate.reactive.panache.Panache // @Deprecated
```

The copied class will be eventually removed



Fixes [this reproducer reported by a user](https://github.com/quarkusio/quarkus/issues/49945)

lucamolteni committed:
Panache Support multiple persistence unit in Hibernate Reactive

* Integration test for multiple reactive persistence units and Panache
Backported from ORM the handling of different persistence units in entity in Panache
* `@WithSessionOnDemand` works only with the default persistence unit
* withSession overload to take the PU name
* Execute update without entity runs on default session
* Repository flush() flushes all PUs
Refactor
* Use ComputingCache instead of a Map<String, LazyValue>
* Keep track of onDemand created sessions to close them accordingly
* create map of entity => PU in kotlin as well (identical to what happens in Hibernate ORM)
* Aligned reactive blocking to orm repositories by using @GenerateBridge to call getSession
* getSession shouldn't be static to be substituted by @GenerateBridge
* Added Panache.withTransaction overload to select persistence unit
* Updated Panache documentation
* Support @WithSession("pu-name") and @WithTransaction("pu-name")
* Added test to verify the same session is used
* Removed dependency from Kotlin panache module to Java, created two new Panache classes inside the Kotlin module and changed the tests accordingly.

FroMage commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/AbstractJpaOperations.java:
```diff
@@ -396,10 +423,12 @@ public static <T extends Mutiny.AbstractQuery> T bindParameters(T query, Map<Str
     }
 
     public static Uni<Integer> executeUpdate(String query, Object... params) {
+        // TODO Luca not sure about this
```
Same thing happens in the ORM version. The day someone wants to support multiple PUs here, we'll ask for what PU to use in `Panache.executeUpdate` where this is called.

FroMage commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java:
```diff
@@ -74,7 +109,8 @@ static <T> Uni<T> withSessionOnDemand(Supplier<Uni<T>> work) {
                 context.removeLocal(SESSION_ON_DEMAND_OPENED_KEY);
-                return closeSession();
+                // TODO Luca this is just wrong
+                return closeSession(DEFAULT_PERSISTENCE_UNIT_NAME);
```
I suppose this should close all currently opened sessions, no?

FroMage replied:
It's debatable.

My opinion is that this says "open a session automatically when asked", and then later (this is lazy), someone opens a session, at which point we know which session and which PU. 

If we only allow a single session open at all times, then this should close that session that we opened, no matter its PU.

If we allow any number of sessions for different PUs open within a `@WithSessionOnDemand` then they should all be closed.

FroMage replied:
I still think this should close the lazily opened session, not the default one.

FroMage commented on extensions/panache/hibernate-reactive-panache/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/Panache.java:
```diff
@@ -26,7 +28,8 @@ public class Panache {
     public static <T> Uni<T> withSession(Supplier<Uni<T>> uniSupplier) {
-        return SessionOperations.withSession(s -> uniSupplier.get());
+        // TODO Luca not sure about this
+        return SessionOperations.withSession(s -> uniSupplier.get(), DEFAULT_PERSISTENCE_UNIT_NAME);
```
It is what it is. We can add an override asking for the PU name.

FroMage commented on extensions/panache/hibernate-reactive-panache/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/PanacheEntityBase.java:
```diff
@@ -102,7 +103,8 @@ public Uni<Void> delete() {
     @JsonIgnore
     public boolean isPersistent() {
-        return INSTANCE.isPersistent(this);
+        // TODO Luca :(
```
I suppose we have to break binary compatibility here 🤷

FroMage commented on extensions/panache/hibernate-reactive-panache/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/PanacheRepositoryBase.java:
```diff
@@ -107,7 +108,9 @@ public default boolean isPersistent(Entity entity) {
     public default Uni<Void> flush() {
-        return INSTANCE.flush();
+        // TODO Luca not sure about this should it flush all the sessions?
+        // TODO Luca this will crash
```
Good question :-/
Yeah, flush all the sessions.

If needed, add an overload for flushing a single PU?

FroMage replied:
Actually… a repository targets a single entity, so it should flush that entity's PU by default, no?

lucamolteni replied:
@FroMage yes but nowhere we save in the repository the type of the entity if I'm not mistaken and it gets erased at runtime. If you think we can find the type of `ENTITY` somewhere can you please me point me how?

lucamolteni replied:
Again, with `@GenerateBridge` everything fits in (many many changes after :))

FroMage commented on extensions/panache/hibernate-reactive-panache/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/PanacheRepositoryBase.java:
```diff
@@ -97,7 +98,7 @@ public default Uni<Void> delete(Entity entity) {
      */
     public default boolean isPersistent(Entity entity) {
-        return INSTANCE.isPersistent(entity);
+        return INSTANCE.isPersistent(entity).await().atMost(Duration.ofSeconds(5));
```
Same, break binary compat here.

FroMage reviewed (changes requested):
Answered inline

yrodiere commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/AbstractJpaOperations.java:
```diff
@@ -67,20 +76,27 @@ public Uni<Void> persist(Stream<?> entities) {
-        return getSession().chain(session -> session.persistAll(entities));
+        Map<String, List<Object>> sessions = Arrays.stream(entities)
+                .collect(Collectors.groupingBy(e -> entityToPersistenceUnit.get(e.getClass().getName())));
+
+        List<Uni<Void>> results = new ArrayList<>();
+        for(Entry<String, List<Object>> entry : sessions.entrySet()) {
+            results.add(getSession(entry.getKey()).chain(session -> session.persistAll(entry.getValue().toArray())));
+        }
+
+        return Uni.combine().all().unis(results).discardItems();
     }
 
     public Uni<Void> delete(Object entity) {
-        return getSession().chain(session -> session.remove(entity));
+        return getSession(entity.getClass()).chain(session -> session.remove(entity));
     }
 
-    public boolean isPersistent(Object entity) {
-        Mutiny.Session current = SessionOperations.getCurrentSession();
-        return current != null ? current.contains(entity) : false;
+    public Uni<Boolean> isPersistent(Object entity) {
+        return getSession(entity.getClass()).map(session -> session.contains(entity));
```
Checking if an entity is persistent does _not_ require a `Uni`: it's a purely local, no-IO operation.

We only need it here because `getSession()` returns an `Uni`. But why does it? Because opening a session in Hibernate Reactive [opens a connection immediately](https://github.com/hibernate/hibernate-reactive/blob/72106db8566eb6504a99e8ef0bf465bcadf0a690/hibernate-reactive-core/src/main/java/org/hibernate/reactive/mutiny/impl/MutinySessionFactoryImpl.java#L91-L93)... Which seems wrong to me: the connection should be opened when first needed, just like in Hibernate ORM.

I will open an issue to Hibernate Reactive.

That being said, I don't think we need to wait for that issue to be fixed. If we need to open a session, we know the instance is not persistent (the session is empty!). So we could do something like `var session = getCurrentSession(entity.getClass()); return session != null && session.contains(entity);`
Does it make sense?

FroMage replied:
The old code did that. You can prove it's not persistent, but you cannot prove it is persistent, though. So I'm not sure I see your point.

FroMage replied:
Oh I see, because if there's a current session, we can query it without having to go through a `Uni`. Duh, sorry 🤦

FroMage replied:
Very good point.

FroMage commented on extensions/panache/hibernate-reactive-panache/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/PanacheRepositoryBase.java:
No, that's not related to stateless sessions. This should be a change done in this PR. The same change was done for the ORM support of multiple PUs.

lucamolteni commented:
setting "ready for review" as I need CI

lucamolteni commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/AbstractJpaOperations.java:
```diff
@@ -67,20 +76,60 @@ public Uni<Void> persist(Stream<?> entities) {
+     * This method flushes *all* persistence units.
+     * Ideally, it should flush only the session associated with the repository’s entity,
+     * but we don’t currently track the entity inside the repository.
+     *
```
used `@GenerateBridge` to get the actual session

lucamolteni commented:
@FroMage thanks to @yrodiere today I managed to do [this](https://github.com/quarkusio/quarkus/pull/50084/commits/4942cbfbfb97898be7b6304a103777ac31fbc2b4) I think it should be what you were thinking about, please take a look

FroMage reviewed (approved):
Perfect! Though, this probably needs to be documented too :)

yrodiere commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java:
```diff
@@ -26,31 +33,44 @@ public final class SessionOperations {
 
-    private static final LazyValue<Key<Mutiny.Session>> SESSION_KEY = new LazyValue<>(
-            new Supplier<Key<Mutiny.Session>>() {
+    private static SessionFactory createSessionFactory(String persistenceunitname) {
```
This could be a function passed to `new ComputingCache(...)`? That way defaults are retrieved implicitly.

yrodiere commented on extensions/panache/hibernate-reactive-panache-common/runtime/src/main/java/io/quarkus/hibernate/reactive/panache/common/runtime/SessionOperations.java:
```diff
@@ -194,26 +250,19 @@ private static Context vertxContext() {
     static void clear() {
-        SESSION_FACTORY.clear();
-        SESSION_KEY.clear();
+        SESSION_FACTORY_MAP.forEachValue(s -> s.close());
```
You shouldn't close session factories. That's done through CDI already.

lucamolteni commented:
> Perfect! Though, this probably needs to be documented too :)

Current state of documentation states:

https://quarkus.io/guides/hibernate-orm-panache#multiple-persistence-units

> Multiple Persistence Units
> 
> The support for multiple persistence units is described in detail in [the Hibernate ORM guide](https://quarkus.io/guides/hibernate-orm#multiple-persistence-units).
> 
> When using Panache, things are simple:
> 
> A given Panache entity can be attached to only a single persistence unit.
> Given that, Panache already provides the necessary plumbing to transparently find the appropriate EntityManager associated to a Panache entity.

So I think they're somewhat fine the way they are now? Nowhere it was written it wasn't working for reactive

lucamolteni commented:
Using the reproducer from the user https://github.com/quarkusio/quarkus/issues/49945 I found another bug that isn't yet covered, so I guess this shouldn't be merged yet

lucamolteni commented:
> And then `Panache.withTransaction(PUType pu, …)` where `PUType` is either `String` if that's how we expose persistence units, or `Class` if that's how.

as DEFAULT_PERSISTENCE_UNIT is a string, I've added an overload with the string [here](https://github.com/quarkusio/quarkus/pull/50084/commits/1fdc218aa84c08e46e9976e7a06ba16a614fec35)

I've also updated the documentation

FroMage reviewed (approved):
Great! Let's get this merged!

yrodiere commented:
Are those failures related? Seems suspiciously close to what is being changed

> #### 📦 integration-tests/hibernate-reactive-panache-kotlin
> 
> ❌ `io.quarkus.it.panache.reactive.kotlin.NoPagingPMT.test` - [History](https://ge.quarkus.io/scans/tests?tests.container=io.quarkus.it.panache.reactive.kotlin.NoPagingPMT&tests.test=test)

lucamolteni commented:
> Yeah, this is weird. But I can't figure out how to find the console output. I guess you'll have to run it locally and pray it fails.

It took me some time to me as well

It's a race condition between 

```kotlin
@Recorder
open class PanacheKotlinReactiveRecorder {
    open fun setEntityToPersistenceUnit(
        entityToPersistenceUnit: Map<String?, String?>?,
        incomplete: Boolean,
    ) {
        AbstractJpaOperations.setEntityToPersistenceUnit(entityToPersistenceUnit)
    }
}
``` 

and

```java
@Recorder
public class PanacheHibernateReactiveRecorder {
    public void setEntityToPersistenceUnit(Map<String, String> entityToPersistenceUnit) {
        AbstractJpaOperations.setEntityToPersistenceUnit(entityToPersistenceUnit);
    }
}
```

Which write in the same field. I'm pretty sure only the Kotlin one should run, but I don't know yet how to do this

yrodiere reviewed (commented):
There's one very suspicious test, but if you're sure of what you're doing...

FroMage commented:
So, do we merge this? I can start rebasing Panache 2 on this then.

lucamolteni commented:
> So, do we merge this? I can start rebasing Panache 2 on this then.

Updated the test and rebased

FroMage commented:
Yay! 🥳
